tests/asserts/equal.cmake

```diff
@@ -1,46 +1,53 @@
 include(cmake_test/cmake_test)
 
-ct_add_test("assert_equal")
+ct_add_test(NAME "assert_equal")
+function(${assert_equal})
     include(cmake_test/asserts/equal)
 
-    ct_add_section("string")
+    ct_add_section(NAME "test_string")
+    macro(${test_string}) #Macro so ${temp} is available to all subsections
```

Comment:
If this is `function` instead of `macro`, then `temp` is not available in the subfunctions? Wouldn't have expected that.

Reply:
That is correct. This is due to the subsections being executed outside of the test function's scope. This could be fixed by adding a `ct_end_test()` before the `endfunction()` which would execute the subsections within the function scope. I had initially decided against this approach as it would add additional boilerplate but now I see the current way just makes things more confusing. I can fix that relatively quickly.

Reply:
I don't understand why they're executed outside of the scope. I assumed that you basically kept running the function until you've exhausted all paths through it. In other words, the first time you call `assert_equal` it registers that there's two sections in there (test_string and test_list) while also running whatever else is in the body of assert_equal. You then run `assert_equal` again, this time when you hit the `ct_add_section` command you call the function provided (`test_string` and `test_list`), which then subsequently add more sections and you rinse and repeat until no new sections are found (I don't think the subsequent redefinition of the function when control returns from a `ct_add_section` would hurt anything).

Reply:
Executing the test function multiple times could have unintended consequences. For example, if you define a target to be used by all subsections, the second execution would redefine the target and result in an unrecoverable error if policy CMP0002 is not set to OLD. We could just tell developers in the documentation to not do this, unless there's someway to undefine all targets/libraries/etc. Do you know of such a way?

Reply:
That's a good point about the targets and no I don't know of any way to undefine a target. Maybe just letting it overwrite each time (with the policy) is okay? If that works, I'd hide the setting of the policy in the main `cmake_test.cmake` file so users don't have to remember to set it. Regardless we'll have to document it.

Reply:
Should be fixed now. I have discovered another minor issue however. When an assertion fails the failure is logged but the test continues running because there isn't currently a way to force the test to return once an assertion fails. This can lead to errors elsewhere if the assertion is expected to force a return as in other lanuages.

For example, if a function uses an assertion to make sure a file exists, the test will fail if it does not exist but the function will continue, possibly using a nonexistant file in later operations. I have not found anything in the CMakePP exceptions documentation that would allow a stack unwind up to the `cpp_catch()` declaration on exception.

The only solution I could come up with is making assertions macros and then calling `return()` inside them when they fail. A `return()` inside a macro will exit the calling function instead of the macro, since macros do not have new scopes. This would cause all subsequent code inside the function after a failed assertion to be skipped, but the scope above that will continue as normal.

Reply:
I think it's fine to crash the test on the first failure (either with an uncaught exception or by calling `ct_exit`) that way it doesn't keep running.

Reply:
I've implemented a `ct_exit()` call when an exception or fatal error is caught while processing a test or section that does not have the `EXPECTFAIL` option, but that still leaves `EXPECTFAIL` tests without any way to mitigate this issue. It's likely not that big of an issue, the only areas I see causing issues are with `file_*` assertions that call other assertions expecting them to unwind on failure. I can fix those assertions myself and put in the documentation that these assertions do not stack unwind or exit on failure when inside an `EXPECTFAIL` test.

Reply:
Why can't you call ct_exit if a function fails to throw?

Reply:
You mean in the case of an `EXPECTFAIL` test that does not throw an exception? Yeah I can add that quick. What I was describing though was that an `EXPECTFAIL` test uses an assert which calls other asserts. The test knows it should fail but the asserts did not and continued using undefined variables, resulting in a low-level `file does not exist` error that partially bypasses my `message()` override (splendid that CMake has multiple disconnected ways of reporting errors where some are not accessible to user code). This caused an internal error state to be set and an error message printed, but because of my override the interpreter did not exit, resulting in additonal errors later and CMake returning an error code. This would cause CTest to report the test as failed even though it actually succeeded.

I have since fixed the asserts so they run additional checks for the files and immediately return when the file is not found. Without some type of stack unwind capability this is probably the best option that I have found.